### PR TITLE
Add Windows support

### PR DESCRIPTION
**Pull request by gitbuda**
- [x] Document `mg_socket`.
- [x] Extend the integration test.
- [x] Implement `mg_socket` for Windows and Linux.

### CODE REVIEW

File: tests/transport.cpp [hunk 1]
```diff
@@ -14,28 +14,31 @@
...
-#include <netinet/ip.h>
-#include <sys/socket.h>
-#include <sys/types.h>
+
+#if MGCLIENT_ON_LINUX
+// NOTE:
+// https://stackoverflow.com/questions/49504648/x509-name-macro-in-c-wont-compile
+#define WIN32_LEAN_AND_MEAN
+#include <openssl/x509.h>
+#endif  // MGCLIENT_ON_WINDOWS
```
Comment (gitbuda) [inline tests/transport.cpp:31]: Fix this

File: CMakeLists.txt [hunk 2]
```diff
@@ -31,10 +50,12 @@ message(STATUS "BUILD_TESTING_INTEGRATION: ${BUILD_TESTING_INTEGRATION}")
 include(CTest)
 
 set(CMAKE_C_STANDARD 11)
+# C++17 is fine here because it is required only for the testing purposes.
 set(CMAKE_CXX_STANDARD 17)
 set(C_STANDARD_REQUIRED ON)
 set(CXX_STANDARD_REQUIRED ON)
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+add_compile_options(-Wall -Wextra -pedantic)
```
Comment (gitbuda) [inline CMakeLists.txt:58]: Remove and make another PR with this + all the fixes.

File: examples/basic.c [hunk 3]
```diff
@@ -9,6 +9,7 @@ int main(int argc, char *argv[]) {
     exit(1);
   }
 
+  mg_init();
```
Comment (antonio2368) [inline examples/basic.c:12]: using C-APi with C++-API looks awkward. Maybe have static wrapper methods in Client?

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -25,6 +25,11 @@ set(mgclient_src_files
         mgsession-encoder.c
         mgtransport.c
         mgvalue.c)
+if(${WIN32})
```
Comment (antonio2368) [inline src/CMakeLists.txt:28]: Why not just WIN32 instead of ${WIN32}?

File: tests/integration/basic_c.cpp [hunk 5]
```diff
@@ -164,4 +166,6 @@ TEST_F(MemgraphConnection, InsertAndRetriveFromMemegraph) {
   ASSERT_EQ(rows, 1);
   ASSERT_EQ(status, 0);
   ASSERT_EQ(mg_session_commit_transaction(session, &result), 0);
+
+  std::cout << "END END" << std::endl;
```
Comment (antonio2368) [inline tests/integration/basic_c.cpp:170]: ?

**Comment by antonio2368** [comment]
Maybe have in src windows/ and posix/ folders for platform specific source files so we don't complicate the names and clutter the src folder?

File: examples/basic.c (hunk 3 above)
Comment (gitbuda) [reply examples/basic.c:12]: Yep, I agree, will do it to be more consistent.

File: src/CMakeLists.txt (hunk 4 above)
Comment (gitbuda) [reply src/CMakeLists.txt:28]: :+1: That's my bash habit, which doesn't make much sense except consistency. But, I think I'll introduce `MGCLIENT_...` CMake variables because otherwise, I would have to repeat CMake specific logic.

File: tests/integration/basic_c.cpp (hunk 5 above)
Comment (gitbuda) [reply tests/integration/basic_c.cpp:170]: Nice catch!

Comment (gitbuda) [reply tests/integration/basic_c.cpp:170]: > Maybe have in src windows/ and posix/ folders for platform specific source files so we don't complicate the names and clutter the src folder?

I was thinking about that, but `mg_socket` should be the only platform-dependent code + there should be 2, maybe 3 implementations (at least in the foreseeable future). I would leave it as is for now and then reevaluate when the support for, e.g., Mac comes in.

**Comment by gitbuda** [comment]
> Maybe have in src windows/ and posix/ folders for platform specific source files so we don't complicate the names and clutter the src folder?

I was thinking about that, but mg_socket should be the only platform-dependent code + there should be 2, maybe 3 implementations (at least in the foreseeable future). I would leave it as is for now and then reevaluate when the support for, e.g., Mac comes in.

File: README.md [hunk 6]
```diff
@@ -53,18 +52,39 @@ make install
...
+ctest
+```
+
+# Building and installing on Windows
+
+To build `mgclient` on Windows MINGW environment could be used.
+   - Install MSYS2 from https://www.msys2.org/.
+   - Install MinGW toolchain with the following command:
+     ```
+     pacman -S --needed git base-devel mingw-w64-i686-toolchain mingw-w64-x86_64-toolchain mingw-w64-i686-cmake mingw-w64-x86_64-cmake mingw-w64-i686-openssl mingw-w64-x86_64-openssl
```
Comment (antonio2368) [inline README.md:68]: pacman?

Comment (gitbuda) [reply README.md:68]: https://www.msys2.org uses `pacman` as a package manager: "_To provide easy installation of packages and a way to keep them updated it features a package management system called Pacman, which should be familiar to Arch Linux users._" :smile:

**Comment by gitbuda** [comment]
> Maybe have in src windows/ and posix/ folders for platform specific source files so we don't complicate the names and clutter the src folder?

I decided to do this because a quick test on Mac showed that it would be nice to have a `mg_socket` implementation for Mac (will come soon), + I agree having folders is cleaner.

File: CMakeLists.txt [hunk 7]
```diff
@@ -20,6 +20,27 @@ project(mgclient VERSION 1.1.0)
 set(mgclient_SOVERSION 2)
 add_definitions(-DMGCLIENT_VERSION="${mgclient_VERSION}")
 
+# Deal with the operating system.
+if(WIN32)
+    message(STATUS "ON WINDOWS BUILD")
+    set(MGCLIENT_ON_WINDOWS TRUE)
+    add_definitions(-DMGCLIENT_ON_WINDOWS=1)
```
Comment (antonio2368) [inline CMakeLists.txt:27]: https://cmake.org/cmake/help/latest/command/add_compile_definitions.html
Is `=1`  necessary? you're only checking with `ifdef`

File: CMakeLists.txt [hunk 8]
```diff
@@ -20,6 +20,27 @@ project(mgclient VERSION 1.1.0)
...
+    # CMAKE_FIND_LIBRARY_PREFIXES on Windows is "lib;".
+    # "lib;" breaks gtest lib referencing.
+    set(MGCLIENT_FIND_LIBRARY_PREFIXES "lib")
+elseif(UNIX AND NOT APPLE)
+    if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+        message(STATUS "ON LINUX BUILD")
+        set(MGCLIENT_ON_LINUX TRUE)
+        add_definitions(-DMGCLIENT_ON_LINUX=1)
```
Comment (antonio2368) [inline CMakeLists.txt:35]: Same as above

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -65,9 +80,8 @@ install(TARGETS mgclient-static mgclient-shared
...
 install(DIRECTORY
         "${CMAKE_SOURCE_DIR}/include/"
-        DESTINATION
-        ${CMAKE_INSTALL_INCLUDEDIR})
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 install(FILES
         "${CMAKE_CURRENT_BINARY_DIR}/mgclient-export.h"
-        DESTINATION
-        ${CMAKE_INSTALL_INCLUDEDIR})
+        "${PROJECT_SOURCE_DIR}/include/mgclient.h"
```
Comment (antonio2368) [inline src/CMakeLists.txt:86]: Isn't this covered with the install above?

File: src/mgclient.c [hunk 10]
```diff
@@ -539,24 +499,29 @@ static int get_hostname_and_ip(const struct sockaddr *peer_addr, char *hostname,
...
-                           NI_MAXHOST, NULL, 0, 0);
-  if (status != 0) {
-    mg_session_set_error(session, "failed to get server name: %s",
-                         gai_strerror(status));
-    return MG_ERROR_NETWORK_FAILURE;
+  // Populate the hostname.
+  // Useful read https://stackoverflow.com/questions/12274028.
+  int nameinfo_status = getnameinfo(peer_addr, sizeof(struct sockaddr),
+                                    hostname, NI_MAXHOST, NULL, 0, 0);
+  if (nameinfo_status != 0) {
```
Comment (antonio2368) [inline src/mgclient.c:510]: What if this fails because of some other reason? You would also pass it as a valid get_hostname_and_ip

File: tests/integration/basic_c.cpp [hunk 11]
```diff
@@ -164,4 +181,35 @@ TEST_F(MemgraphConnection, InsertAndRetriveFromMemegraph) {
   ASSERT_EQ(rows, 1);
   ASSERT_EQ(status, 0);
   ASSERT_EQ(mg_session_commit_transaction(session, &result), 0);
+
+  {
+    ASSERT_EQ(mg_session_run(session,
```
Comment (antonio2368) [inline tests/integration/basic_c.cpp:186]: Isn't this for unit test and not for integration/e2e?

File: tests/integration/basic_cpp.cpp [hunk 12]
```diff
@@ -113,4 +120,31 @@ TEST_F(MemgraphConnection, InsertAndRetrieveFromMemgraph) {
   }
   ASSERT_EQ(result_counter, 1);
   ASSERT_TRUE(client->CommitTransaction());
+
+  {
+    ASSERT_TRUE(client->Execute(
```
Comment (antonio2368) [inline tests/integration/basic_cpp.cpp:125]: Same as above

File: tests/unit/mgclient_value.cpp [hunk 13]
```diff
@@ -29,7 +32,7 @@ TEST(ValueTest, BasicTypes) {
   Value value_bool1(true);
   Value value_bool2(false);
   Value value_int1(-13);
-  Value value_int2(1L << 60);
+  Value value_int2((int64_t)(1L << 60));
```
Comment (antonio2368) [inline tests/unit/mgclient_value.cpp:35]: static_cast

File: src/CMakeLists.txt (hunk 9 above)
Comment (gitbuda) [reply src/CMakeLists.txt:86]: Yea, I don't remember why I did this... will check and fix...

File: src/mgclient.c (hunk 10 above)
Comment (gitbuda) [reply src/mgclient.c:510]: The hostname defined here is used in the SSL related user-defined trust callback. The hostname will be passed to the user code once the callback is called. In case of any failure, the user will get `unknown` and has to decide what to do (user code has to deal with that situation). Previous code is problematic because the whole client doesn't work because that `getnameinfo` failed, even if SSL is NOT used (it failed in a trivial case when `peer_addr` was constructed from `localhost` on Windows, a common case). I was thinking of adding a warning, but it doesn't make sense because in the trivial case user will get a useless alert (he doesn't deal with the callback at all). I see two options:
1) I have to figure out what is going on with these two calls, `getaddrinfo` and `getnameinfo`. That's all connected and quite complex because it's platform dependent.
2) Explain to the user what is going on in case he gets `unknown` + a guide on how to handle it.

I would prefer option two at this point.

File: tests/integration/basic_c.cpp (hunk 11 above)
Comment (gitbuda) [reply tests/integration/basic_c.cpp:186]: It is, but I would like to have something like this (probably more, with more stuff tested) on the integration/end2end level because these tests will be called on each platform. It's going to be required that all integration tests pass on each supported platform. Not all unit tests work on every platform (we have to port everything which will take more time / it's not critical at this point, e.g., socketpair is tricky ON_WINDOWS). I mean, I can move this in the `tests/unit` but I would like to keep this test as it is now (or very similar).

File: src/mgclient.c (hunk 10 above)
Comment (gitbuda) [reply src/mgclient.c:510]: The third option is to combine. `get_hostname_and_ip` can be placed under `MG_SSLMODE_REQUIRE` + in the failure case, a warning can be logged + the hostname can be something equivalent to `unknown`?

File: tests/integration/basic_c.cpp (hunk 11 above)
Comment (antonio2368) [reply tests/integration/basic_c.cpp:186]: Oh yeah, forgot we won't have unit tests on all platforms for now.
Could you just add a comment why this isn't in unit folder then?

File: src/mgclient.c (hunk 10 above)
Comment (antonio2368) [reply src/mgclient.c:510]: Hmm, I agree that it doesn't make sense to LOG it to the user as he can't do much about it if I understood it correctly. But adding DLOG would be great to have!

**Review by antonio2368 — APPROVED** [review]
